Usable gtm instance in composition API

Abdulrahmanh995 opened the pull request:
Allow using `gtm` instance in `setup` method by injecting the object inside.

```
setup() {
      const gtm = inject('gtmInstance');
      console.log('gtm', gtm);
}
```

Abdulrahmanh995 committed:
Update plugin.ts

Move on extra information to the bottom of dataLayer.push call in order to allow overriding event name and other properties.

Abdulrahmanh995 committed:
Update index.ts

Abdulrahmanh995 committed:
Update plugin.ts

Shinigami92 commented:
This is a Vue 3 only feature, isn't it?

Shinigami92 commented:
Also I'm currently for a few more days in vacation, so I will be back in a few days

Abdulrahmanh995 commented:
Yes, it's Vue 3 related

Shinigami92 commented on src/index.ts:
```diff
@@ -42,7 +42,9 @@ function install(Vue: App, initConf: VueGtmUseOptions = { id: "" }): void {
-  Vue.config.globalProperties.$gtm = new GtmPlugin(pluginConfig.id);
+  const gtmInstance = new GtmPlugin(pluginConfig.id);
+  Vue.config.globalProperties.$gtm = gtmInstance;
+  Vue.provide("gtmInstance", gtmInstance);
```
About the `"gtmInstance"` naming, is there any convention or other plugins that also using `xInstance`?
Otherwise, what was your idea behind the `Instance` suffix? Is there a better name like `GTM`, `gtmPlugin`, `gtmPluginInstance`, ...?

Also we should maybe add a section to the README so others can also benefit from it in their `setup` methods?

Shinigami92 commented:
@Abdulrahmanh95 oh and maybe we can provide the string as a exported constanct, so others can import it like

```ts
import { GTM_INJECT_NAME } from 'vue-gtm'

setup() {
  const gtm = inject(GTM_INJECT_NAME)
}
```

What do you think?

Shinigami92 reviewed (changes requested):
@Abdulrahmanh95 I also updated master, so if you merge it into your branch, this may solve the outdated deps test step

ahashem95 committed:
- Rename provided gtm object to gtmPlugin.
- Update README.md with how to use composition API.
- Add a custom hook to access the gtm object.

Abdulrahmanh995 committed:
Merge pull request #1 from mib200/master

Update devDependencies

ahashem95 committed:
Merge remote-tracking branch 'origin/master'

ahashem95 committed:
- Define hook inside index.ts instead of utils.ts.

ahashem95 committed:
- Get rid of providing the gtm instance. Depend on a variable instead.

Abdulrahmanh995 commented:
What about using a custom composable (hook) so the developer will only call `useGtm()` to access the plugin with composition API.
Please recheck the new changes I've added. I also updated `readme.md`

ahashem95 committed:
- Get rid of unused constants and imports.

ahashem95 committed:
- Get rid of unused constants and imports.

Shinigami92 commented on README.md:
```diff
@@ -185,6 +185,27 @@ const myRoute = {
+
+const gtm = useGtm();
+
+gtm.trackEvent({
```
Because this is (I think) especially for a `setup` method, should we create an example including this `setup` method so it's more real-world example?

Abdulrahmanh995 replied:
Yes sure. I updated the example to be more detailed.

Shinigami92 commented on README.md:
````diff
@@ -185,6 +185,27 @@ const myRoute = {
+```js
+import {useGtm} from 'vue-gtm';
+
+const gtm = useGtm();
````
I'm a little bit confused how this is working.
We have a kinda global variable `gtmPlugin` in line 6, this is initialized in the `install` function.
So first: if you don't call this `install` function it will never be initialized :warning: (?!). So it is more like `let gtmPlugin: GtmPlugin | undefined;` and `export function useGtm(): GtmPlugin | undefined {`. Or am I so wrong and miss anything?
Do other plugins also use something like this `use*` function? I know `createRouter` from https://next.router.vuejs.org/guide/migration/#new-router-becomes-createrouter but I'm completely not into Vue 3 at all cause I'm blocked by Vuetify not supporting Vue 3 right now :disappointed: 

Sorry for all these questions but I just want to learn and make it right in the way that it is concise with other plugins out there so we don't introduce multiple standards
![image](https://imgs.xkcd.com/comics/standards.png)

Abdulrahmanh995 replied:
No worries. The way of Vue3 composition API has completely changed the game by putting all of your code inside `setup` method.

But as the code may get bigger and bigger, it's not logical to put hundreds of lines inside this single method (`setup`).

So if you check this [link ](https://v3.vuejs.org/guide/composition-api-introduction.html#standalone-computed-properties) about composition API and how to define a `custom reusable composable` which is simply a reusable block of logic defined inside a function, you can understand why I created this `useGtm` function.

You can notice that examples in the docs contain functions that use naming convention of `use*`.
However all other plugins like `vue-router@next` has its own composable like `useRouter` or `useRoute` and `vue-store` also has `useStore` so the user has access to those objects with the need of using `this`.

Also, inside your own app you should define your own composables like `useStorage` or `useHttp` or etc...

So using this way, I'm not confusing the user about injection and using a custom string of constant and I'm not developing my own standard but I'm just wrapping this local variable `gtmPlugin` with a composable that is exported to be used just like other libraries or plugins.

For your point of initialization, you are totally true. If the user didn't register the plugin so it won't be simply initialized because as I guess, the initialization step is mandatory as the [README.md](https://github.com/mib200/vue-gtm#configuration) says.

Shinigami92 replied:
Brilliant, that's exactly what I needed.
(virtually marked as resolved, but I will let it open so it stays open)

ahashem95 committed:
- Change the return type of useGtm and type of gtmPlugin.

ahashem95 committed:
- Update README.md example to be more detailed.

Shinigami92 commented:
So theoretically this would get an **approve** from me.
But I want to know two more things:

1. Looking into this code:  https://github.com/vuejs/vuex/blob/1c693b6c917d1420032ad2c95680ce38ff2e6bff/dist/vuex.esm-browser.prod.js#L10-L14
   It seems that they use `inject` inside the `use*` function. Do you thing this could make it even better?
2. Sadly I don't have any kind of tests for this repo. Did you test it (e.g. with `yarn link`, and I think potentially remove the inner node_modules folder out of the linked module. Or maybe just modify the dist of the regular dependency and test it this way).
   What you need to test is if you get the instance out of it in another file. So you have the `Vue.use` (or `createGtm`) in the `main.ts` and then you have another file in your Vue project were `useGtm` is called and it doesn't return `undefined` but the instance.
   Just asking cause I'm afraid how _import_ strategy can potentially just call `vue-gtm`'s `index` twice and then just don't call the `install` function at all. _hope you understand what I mean_

If this works, I'm very happy to merge it.

Apart from that, do you thing this could also work for the `vue-2` branch? Or should I just don't merge this into `vue-2` branch?

Abdulrahmanh995 commented:
As I know you can't use `inject` outside `setup` method in Vue 3, check this [link](https://v3.vuejs.org/api/composition-api.html#provide-inject), so I think using our way is simple enough to get the work done.

For the second point, yes sure, I've tested it with the website I'm currently working on [Timeviewer](https://timeviewer.tv) 😄  using my github master branch.

Calling `createGtm` inside `main.js` and using `useGtm` inside `setup` function of my components is totally fine 👍 and working well in production now 🚀.

For `Vue-2` branch, it will totally work fine as the changes and `useGtm` function don't affect anything.
